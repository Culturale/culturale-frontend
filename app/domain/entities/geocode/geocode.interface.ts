/**
 * Representation of a geographical point
 *
 * @property latitude
 * @property longitude
 */
export interface IGeoCode {
  latitude: number;
  longitude: number;

  distanceToLocation: (location: IGeoCode) => number;
}
