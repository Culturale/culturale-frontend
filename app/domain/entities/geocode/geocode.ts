import type { IGeoCode } from './geocode.interface';

const EARTH_RADIUS_METERS = 6371000;

function toRadians(value: number): number {
  return (value * Math.PI) / 180;
}

export class Geocode implements IGeoCode {
  public readonly latitude: number;
  public readonly longitude: number;

  constructor(longitude: number, latitude: number) {
    this.latitude = latitude;
    this.longitude = longitude;
  }

  public distanceToLocation(location: IGeoCode): number {
    const dLat = toRadians(location.latitude - this.latitude);
    const dLon = toRadians(location.longitude - this.longitude);
    const lat1 = toRadians(this.latitude);
    const lat2 = toRadians(location.latitude);

    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.sin(dLon / 2) * Math.sin(dLon / 2) * Math.cos(lat1) * Math.cos(lat2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const d = EARTH_RADIUS_METERS * c;
    return d;
  }
}
