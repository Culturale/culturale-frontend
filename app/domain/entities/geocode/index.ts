export * from './geocode';
export * from './geocode.interface';
