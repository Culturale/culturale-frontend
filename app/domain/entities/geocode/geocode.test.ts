import { Geocode } from './geocode';
import type { IGeoCode } from './geocode.interface';

describe('Geocode', function () {
  let instance: IGeoCode;

  beforeEach(function () {
    instance = new Geocode(42, 2);
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });

  describe('distanceToLocation', function () {
    it('returns the correct distance', function () {
      expect(instance.distanceToLocation(new Geocode(38, 1))).toBe(458315.0183091454);
    });
  });
});
