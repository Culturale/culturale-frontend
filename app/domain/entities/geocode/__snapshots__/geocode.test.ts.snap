// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geocode can be created 1`] = `
Geocode {
  "latitude": 2,
  "longitude": 42,
}
`;
